### PR TITLE
PopupMenuButton: create IconButton if child is Icon

### PR DESCRIPTION
**Pull request by perlatus**
Otherwise the resulting button has an abnormally small and rectangular area. With multiple PopupMenuButton(child: Icon) they get squished together in the AppBar.

### CODE REVIEW

**Comment by Hixie** [comment]
Can you elaborate on what effect are you trying to achieve and what led to this change?

**Comment by perlatus** [comment]
Sure, here are some screenshots.
If you have an AppBar and fill it with PopupMenuButton(child: Icon), you see the following:
icon
icon-overflow

If you make them PopupMenuButton(child: IconButton), you get:
hack
hack-overflow

With this fix, all of the buttons are homogenous:
fix

**Comment by Hixie** [comment]
Ah, I see.

I don't think this is the right fix. Consider for example if someone wanted to use an `ImageIcon` or a `MySpecialVectorIcon` or whatever.

Maybe instead we should have a new property `icon` on the `PopupMenuButton`, and either the `icon` or the `widget` (or both) should be null, and if the `icon` is non-null then it gets used with an `IconButton`, otherwise the `child` gets used in the `InkWell` like today?

Or we could have a property that controls how `child` is used, in `IconButton` or in `InkWell`.

**Comment by perlatus** [comment]
I like the idea of having a separate `icon` argument since it's a common use-case.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 1]
```diff
@@ -482,8 +487,10 @@ class PopupMenuButton<T> extends StatefulWidget {
     this.tooltip: 'Show menu',
     this.elevation: 8.0,
     this.padding: const EdgeInsets.all(8.0),
-    this.child
+    this.child,
+    this.icon
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: nit: add a trailing comma

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 2]
```diff
@@ -512,6 +519,9 @@ class PopupMenuButton<T> extends StatefulWidget {
   /// The widget below this widget in the tree.
   final Widget child;
 
+  /// The icon used for this button.
+  final Icon icon;
+
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: we should add some more depth to the docs for `child` and `icon` that explains what each does and how it relates to the other and that `icon` has a default if neither is provided.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 3]
```diff
@@ -554,17 +564,14 @@ class _PopupMenuButtonState<T> extends State<PopupMenuButton<T>> {
...
-    );
+    return widget.child != null
+      ? new InkWell(
+          onTap: showButtonMenu,
+          child: widget.child)
+      : new IconButton(
+          icon: widget.icon ?? _getIcon(Theme.of(context).platform),
+          padding: widget.padding,
+          tooltip: widget.tooltip,
+          onPressed: showButtonMenu);
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: you need some extra newlines (before the `)`s) and some trailing commas (on those last lines) to match the prevailing style.

**Comment by Hixie** [comment]
This will need tests also. :-)

File: packages/flutter/lib/src/material/popup_menu.dart (hunk 2 above)
Comment (perlatus) [inline packages/flutter/lib/src/material/popup_menu.dart]: I described that [here with the class-level docs](https://github.com/flutter/flutter/pull/10230/commits/30b085b6ca19974259b878bf0f04c22e7839de49#diff-437fce7f7d560afa95e469f92c2c8808R473).

**Comment by perlatus** [comment]
I can't pin down the analyzer errors:
```
||     error • In constant expressions, operands of this operator must be of type 'bool', 'num', 'String' or 'null' at lib/src/material/popup_menu.dart:493:15 • const_eval_type_bool_num_string
||     error • In constant expressions, operands of this operator must be of type 'bool', 'num', 'String' or 'null' at lib/src/material/popup_menu.dart:493:32 • const_eval_type_bool_num_string
```

Those are for the usage of `child` and `icon` inside the assert, but it doesn't make sense to me.

**Comment by Hixie** [comment]
Those two analyzer errors are a known bug. We're going to fix it Monday hopefully with a Dart roll. Just disregard it for now.

**Comment by Hixie** [comment]
Let me know (here or on gitter) if you need any help writing tests for this. Thanks!

**Comment by perlatus** [comment]
I'm still watching these PRs; just haven't had time to work on them.

**Comment by Hixie** [comment]
Cool, no rush!

**Comment by Hixie** [comment]
@perlatus Please reopen this PR (or comment on it asking me to reopen, I'm not sure if you are able to reopen PRs) when you are ready to land it. Thanks! (I'm just cleaning out our review queue and this one has been open with no activity for more than a week.)

**Comment by perlatus** [comment]
@Hixie I updated this with the tests. Still on https://github.com/perlatus/flutter/tree/popupmenu-iconchild

**Comment by perlatus** [comment]
@Hixie can you reopen this PR?

**Comment by Hixie** [comment]
done!

**Comment by perlatus** [comment]
@Hixie this is ready to go, assuming that those analyzer errors aren't my fault.

File: packages/flutter/lib/src/material/popup_menu.dart [hunk 4]
```diff
@@ -672,8 +677,10 @@ class PopupMenuButton<T> extends StatefulWidget {
     this.tooltip: 'Show menu',
     this.elevation: 8.0,
     this.padding: const EdgeInsets.all(8.0),
-    this.child
+    this.child,
+    this.icon,
   }) : assert(itemBuilder != null),
+       assert(child == null || icon == null), // or both
```
Comment (Hixie) [inline packages/flutter/lib/src/material/popup_menu.dart]: you're hitting an analyzer bug: https://github.com/dart-lang/sdk/issues/30288
change the `||` to `&&`, the `==` to `!=`, and prefix the whole thing with a `!` and it'll mean the same thing but work.
`assert(!(child != null && icon != null))`

**Comment by Hixie** [comment]
LGTM

**Comment by Hixie** [comment]
Thanks for your contribution! I shall now land this.

**Pull request merged**